plugins/modules/dcnm_image_upgrade.py

```diff
@@ -0,0 +1,5313 @@
+        super().__init__(module)
+        self.class_name = self.__class__.__name__
+        self._init_properties()
+
+    def _init_properties(self):
+        self.properties = {}
+        self.properties["epld"] = False
+        self.properties["issu"] = True
+        self.properties["ndfc_data"] = None
+        self.properties["ndfc_response"] = None
+        self.properties["ndfc_result"] = None
+        self.properties["package_install"] = False
+        self.properties["policy_name"] = None
+        self.properties["serial_number"] = None
+        self.properties["epld_modules"] = None
+
+    def refresh(self):
+        """
+        Refresh self.data with current install-options from NDFC
+        """
+        if self.policy_name is None:
+            msg = f"{self.class_name}.refresh: "
+            msg += "instance.policy_name must be set before "
+            msg += "calling refresh()"
+            self.module.fail_json(msg)
+        if self.serial_number is None:
+            msg = f"{self.class_name}.refresh: "
+            msg += f"instance.serial_number must be set before "
+            msg += f"calling refresh()"
+            self.module.fail_json(msg)
+
+        path = self.endpoints.install_options.get("path")
+        verb = self.endpoints.install_options.get("verb")
+        self._build_payload()
+        msg = f"REMOVE: {self.class_name}.refresh: "
+        msg += f"payload: {self.payload}"
+        self.log_msg(msg)
+        self.properties["ndfc_response"] = dcnm_send(
+            self.module, verb, path, data=json.dumps(self.payload)
+        )
+        msg = f"REMOVE: {self.class_name}.refresh: "
+        msg += f"ndfc_response: {self.ndfc_response}"
+        self.log_msg(msg)
+        self.properties["ndfc_result"] = self._handle_response(self.ndfc_response, verb)
+        # TODO:2 should error message contain full response or just DATA.error?
+        if self.ndfc_result["success"] is False:
+            msg = f"{self.class_name}.refresh: "
+            msg += "Bad result when retrieving install-options from NDFC. "
+            msg += f"NDFC response: {self.ndfc_response}"
+            self.module.fail_json(msg)
+
+        self.properties["ndfc_data"] = self.ndfc_response.get("DATA")
+        self.data = self.properties["ndfc_data"]
+        if self.data.get("compatibilityStatusList") is None:
+            self.compatibility_status = {}
+        else:
+            self.compatibility_status = self.data.get("compatibilityStatusList")[0]
+
+    def _build_payload(self):
+        """
+        {
+            "devices": [
+                {
+                    "serialNumber": "FDO211218HH",
+                    "policyName": "NR1F"
+                }
+            ],
+            "issu": true,
+            "epld": false,
+            "packageInstall": false
+        }
+        """
+        self.payload = {}
+        self.payload["devices"] = []
+        devices = {}
+        devices["serialNumber"] = self.serial_number
```

Comment:
Why is it self always has a single serial number and policy name. We only support a single image upgrade to a single switch at a time?

Reply:
We support multiple switches at a time and the code does as well.  It's only this class that supports one at a time.  Later, the individual switches are all merged into a single payload that is sent to NDFC.